### PR TITLE
sdk: Extract signature crate

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
`solana_sdk::signature::Signature` needs to be made available outside of solana-sdk.

#### Summary of Changes
- Make a new solana-signature crate and move `Signature`, `SIGNATURE_BYTES`, and `ParseSignatureError` there
- Re-export these items in `solana_sdk::signature` for backwards compatibility
- Make the `serde*`, `rand` and `ed25519-dalek` dependencies optional in this crate. There a bunch of cases where we want `Signature` but don't want all these extras

### CODE REVIEW

**Review by buffalojoec — COMMENTED** [review]
@kevinheavey This type is not in the program SDK. Do we need to do this one now? I was thinking we could bust up the pieces in the program SDK according to https://github.com/anza-xyz/agave/pull/2569 first.

cc @joncinque

**Comment by kevinheavey** [comment]
Why wait? Splitting up the program crate won't affect this PR

**Comment by joncinque** [comment]
Even though it's outside the program sdk, it's a really obvious candidate for splitting out of the non-program SDK, kind of like Hash, since it's just encapsulating 64 bytes. If I'm not mistaken, we've even had Agave devs want this, so I'll be happy to take a look through it

File: sdk/signature/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
+//! 64-byte signature type.
+#![cfg_attr(RUSTC_WITH_SPECIALIZATION, feature(min_specialization))]
+#[cfg(feature = "serde")]
+use serde_derive::{Deserialize, Serialize};
+#[cfg(any(test, feature = "verify"))]
+use std::convert::TryInto;
+use {
+    core::str::from_utf8,
+    generic_array::{typenum::U64, GenericArray},
+    std::{fmt, str::FromStr},
```
Comment (joncinque) [inline sdk/signature/src/lib.rs]: what do you think about adding an `std` feature to this crate too and making it no-std? It looks like just the `TryFrom<Vec<u8>>` and `Error` implementations require it

File: sdk/Cargo.toml [hunk 2]
```diff
@@ -23,11 +23,11 @@ default = [
 full = [
     "byteorder",
     "chrono",
-    "generic-array",
```
Comment (joncinque) [inline sdk/Cargo.toml:26]: :partying_face:

**Review by joncinque — COMMENTED** [review]
This is looking good. Just the one suggestion about whether we can make this no-std, and after that, this should be good to go!

**Comment by kevinheavey** [comment]
I noticed we could swap out GenericArray but I've made a separate PR for that in case it turns into a rabbit hole https://github.com/anza-xyz/agave/pull/2857

**Review by joncinque — APPROVED** [review]
Looks great, thanks for the contribution! I also tested this locally with `--no-default-features` to make sure that the features on the sdk and new crate are all good.

**Comment by joncinque** [comment]
> I noticed we could swap out GenericArray but I've made a separate PR for that in case it turns into a rabbit hole https://github.com/anza-xyz/agave/pull/2857

I deliberately didn't mention it for exactly that reason :sweat_smile:

**Pull request merged**